### ISSUE TITLE
Continous Auto Trim active during maneuvers

### ISSUE THREAD

**Issue by b14ckyy** [iNavFlight/inav#9912]
## Current Behavior
If Continuous Servo Trim is enabled on fixed wings, it is constantly trimming even when not flying level or when flying slow or fast maneuvers. This leads to a few seconds of strongly out of trim plane behavior after long rolls or turns as well as loops and can in worst case cause a crash at a low altitude loop. It also causes a lot of false trim during landing when the pilot does a flare maneuver at low airspeed. 

## Steps to Reproduce
1. Enable Continous Auto Trim in Features
2. Fly in acro mode and fly level until plane flies dead straight with no stick input
3. Do a long continous bank 180° or more turn or fly a spiral
4. quickly level off and release the sticks
Result: the plane will be heavily out of trim so that a low I-Term cannot compensate anymore and it takes a few seconds to recover

## Expected behavior
Continous Servo Trim should only trim the plane if the rate setpoint is very close to 0 in Autolevel modes or within stick deadband in acro mode. The trim should be locked while the pilot or autopilot is doing any maneuvers, turns or altitude changes. 

## Suggested solution(s)
I checked the code as far as I was able to understand it and found that the Continous Servo trim has multiple sanity checks before it should get active. 

```
planeIsFlyingStraight && 
noRotationCommanded && 
planeIsFlyingLevel &&
sticksAreCentered &&
!FLIGHT_MODE(MANUAL_MODE) && 
isGPSHeadingValid() // TODO: proper flying detection
```
I think one or more of these sanity checks are not properly working. 

## Additional context
The issue must be in INAV for a long time. I just noticed it intensively while testing other stuff and investigated a bit more. Here are 3 flight logs I did for a test of https://github.com/iNavFlight/inav/pull/9905 on a 8.0 build: [LOG_2.zip](https://github.com/iNavFlight/inav/files/14916408/LOG_2.zip)

LOG 33 is with Continous servo trim enabled. You clearly see after a long turn, that Integral is pushing hard to keep the pitch rate at 0 and settles slowly as trim is corrected. 
LOG 34 I disabled CST and did a classic autotrim with the trim mode, clearly does not show this same 
LOG 35 is another flight after saving the trim and just redoing autotune again to validate the correct rates and FF with a persistent trim. 

---

- FC Board name and vendor: Matek F405-WTE
- INAV version string: INAV/MATEKF405TE_SD 8.0.0 Apr  8 2024 / 18:01:53 (55a16f67)

**Comment by Yury-MonZon** [comment]
Few lines above the mentioned if statement we have:
````

**Comment by ProfDrYoMan** [comment]
I did ask the same in the mentioned other ticket:

Any news on this?

Would love to keep servo auto trim enabled all the to ensure proper trim even if CG is moved or motor angle was adjusted.

This is a neat feature, but the flaw does prevent me from enabling it.

**Comment by sensei-hacker** [comment]
This might do the trick, if someone wants to test it
https://github.com/iNavFlight/inav/pull/11215

**Comment by b14ckyy** [comment]
@sensei-hacker  I will try to give it a shot when I can but not sure if I can do it within the next 2 weeks. 
Btw are you sure its just just I-Term translating after ending a maneuver? Because in that case, with the new I-Term lock feature of 8.0, shouldn't that already prevent it since I-Term is attenuated to 0 and will only recover when the rate-error is below 10°/s.

**Comment by sensei-hacker** [comment]
That's a good question.

I-term lock reduces pidState->errorGyroIf (also known as getAxisITerm) when the conditions are met.

I see what auto trim does is it adds a portion of I-term to the servo midpoint:

https://github.com/iNavFlight/inav/pull/11215/changes#diff-c90491cca5be9a11d702c750699464d7fbc1cc617e32cd982070552522a70207L655

From this, I suppose we can conclude that whatever is happening with auto trim, it's happening when the conditions are not such that I-term lock would set the I-term to zero. I guess the question is, is it during conditions such that I-term is *changing* significantly? That's what this change will test / handle.

**Comment by b14ckyy** [comment]
From my experience, the trim-error usually happens after a long lasting input and is especially noticeable on pitch. The best way to replicate it is, by flying in a high 80°+ bank turn for a full circle or at least for a few seconds continuously and then leveling off and releasing the stick. All in Acro mode. Depending on the tune, the plane might then pitch down or up. 

if FF is a bit too high and I-Term was fighting against your stick input to correct for that and hit the target rate, the plane will pitch down after leveling off. If the FF was too low the trim will go up. 

I never had issues for short sharp inputs but only after longer constant inputs. As if the trim is still active with setpoint !=0 although it should be inactive. Its hard to tell from the logs though. 

But anyway. My suggestion is that we don't put this fix into 9.0 until we have fully confirmed. So far this behavior is more an annoyance in rare conditions than a big issue (although it might have caused one crash for me by counteracting my pull up from a low dive, but not sure). Since I know exactly what planes in my hangar show the behavior the most, I will do some proper testing with and without your change, and if it helps, we can still add it to 9.x or 9.0.x later.

**Comment by sensei-hacker** [comment]
Oh for sure. 9.0 is pretty much locked down at this point. 
This would be for 9.1

**Comment by Carticiak** [comment]
What about including whether or not flaps/airbrakes are engaged? Having a user-set rc channel determine whether or not it's active makes most sense to me. 

For example, I set ch 11 as my "air brake" channel. When the switch is in its lowest position and the airbrakes are closed, it passes the sanity check.

I have also never tried using the flap user mode before, so this might be also doing exactly what I describe.

**Comment by sensei-hacker** [comment]
> Having a user-set rc channel determine whether or not it's active makes most sense to me.

One might ask, how that much different from regular auto trim?

**Comment by Carticiak** [comment]
> > Having a user-set rc channel determine whether or not it's active makes most sense to me.
> 
> One might ask, how that much different from regular auto trim?

Including flap/airbrake engagement in the list of sanity checks for the autotrim to engage is what I was getting at.

**Comment by b14ckyy** [comment]
With flaps and airbrakes, you are supposed to have a pitch mix implemented yourself that changes the pitch mixer output when deployed. This should not rely on Integral or Auto trim since that will always have a delayed response.

**Comment by MrD-RC** [comment]
True, you should set up elevator compensation to deal with flap/flaperon/spoileron/airbrake deployment. This would be configured with continuous servo trim disabled. That would be the correct way to do it. 

But it's worth noting that it is a huge pain to set up and tune. As you can't just use trims on the radio. There's a lot of flying, enabling flaps seeing the results, landing, adjust the mixer, and repeating. Also the built in flaperon mode doesn't have this. It could be added through a mixer. But would still be at least as big a pain in the ass to tune. No doubt most people don't add compensation. 

It would be nice if certain actions could disable continuous servo trim. For example lowering landing gear (not just the flap type operations). But, the trick is what conditions cause that to be wanted. This is why I think it would be better in the IPF. So if RC 11 > 1050 disable the continuous servo trim. Or if RC 8 LOW disable the continuous servo trim. That would be pretty simple to implement.

**Comment by MrD-RC** [comment]
> > Having a user-set rc channel determine whether or not it's active makes most sense to me.
> 
> One might ask, how that much different from regular auto trim?

Very different.

One, you want the trims to update all the time. You're just specifying conditions when this actually should be paused. So this is more about the auto trim is always happening, and you're just flying normally. You aren't explicitly saying to no longer trim. It's just that certain profiles of the flight stop it. 

The other (servo auto trim mode), you are enabling when the aircraft is to be trimmed. It is a deliberate tuning activity. Also, from what I remember, you're supposed to keep that active until you land and disarm. So there's nothing to say that the flaps/gear/etc wouldn't mess with that trim.

**Comment by b14ckyy** [comment]
@MrD-RC 

> But it's worth noting that it is a huge pain to set up and tune. As you can't just use trims on the radio. 

Nope. Just one flight. That's why we have blackbox. Can even use OSD values. 
Fly level, make sure plane is trimmed, no elevator mix. Switch manual mode. Deploy flaps and counteract with your pitch stick. hold it so the plane flies level again and you will have the exact offset value in the log or your OSD. 

Just add the mixer then and change the percentage with deployed flaps on the ground until the servo output for elevator matches what seen in the test flight. Done.

**Comment by MrD-RC** [comment]
Pah, who uses blackbox 😉

I still think having disable conditions would be handy. Landing gear for one.

**Comment by b14ckyy** [comment]
Or alternatively, we don't save it? Reset to the pre flight value on disarm? 
Most actual issues arise from a bad trim after flare. 

instead use the auto trim mode once for a baseline and then just don't change the actual midpoints anymore by continous servo trim.

**Comment by MrD-RC** [comment]
Or keep updating the values. But keep 2 sets. The current live value. Plus the live value from 2 minutes ago. Once landing is detected. Use the 2 minutes old value to save. 

If there is no 2 minute value (or the flight time is less than say 5 mins) don't save.

**Comment by b14ckyy** [comment]
also an option. Lets fix the stick input trim first and see if that helps. Further refinement can then be done.

**Comment by sensei-hacker** [comment]
> You aren't explicitly saying to no longer trim. It's just that certain profiles of the flight stop it.
> 
> The other (servo auto trim mode), you are enabling when the aircraft is to be trimmed. It is a deliberate tuning activity. Also, from what I remember, you're supposed to keep that active until you land and disarm.


FYI I was responding to "Having a user-set rc channel determine whether or not it's active makes most sense to me."

Anyway, the PR has been merged that covers what is probably the most common case.
